File: 0121-买卖股票的最佳时机/Solution.java
class Solution {
    public int maxProfit(int[] prices) {
        if(prices == null || prices.length < 2){
            return 0;
        }
        // 谷底, 即买入的最低价
        int minprice = prices[0];
        int maxvalue = 0;
        // 遍历每一天价格的时候，
        // 要么prices[i]小于minprice, 则更新minprice
        // 要么prices[i]大于minprice, 则选择更新maxvalue 
        for(int i=1; i<prices.length; i++){
            if(prices[i] < minprice){
                minprice = prices[i];
            }else if(prices[i]-minprice > maxvalue){
                maxvalue = prices[i]-minprice;
            }
        }
        return maxvalue;
    }
}

File: 0121-买卖股票的最佳时机/solution.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        ret = 0
        if len(prices) < 2:
            return ret
        dp = [0] * len(prices)
        for i in range(1, len(dp)):
            dp[i] = max(dp[i-1], prices[i]-min(prices[:i]))
            ret = max(ret, dp[i])
        return ret

    def maxProfit(self, prices: List[int]) -> int:
        ret = 0
        buy = float("inf")
        for i, price in enumerate(prices):
            if price <= buy:
                buy = price
            else:
                ret = max(ret, price - buy)
        return ret
                
